lib/middleware.ts

```diff
@@ -113,6 +113,44 @@ var defaultHandlers = [
   () => ({title: "request-finished"}),
 ];
 
+const defaultContextHandlers = [
+  (req) => ({"request-id": req.headers["x-request-id"]}),
+  (req) => ({"session-id": req.sessionID}),
+];
+
+function handlerData(handlers, ...args) {
+  const data = {};
+  handlers.forEach((h) => {
+    try {
+      const handler_data = h(...args);
+      _.extend(data, handler_data);
+    } catch (e) {
+      // ignore invalid handler
+    }
+  });
+  return data;
+}
+
+class ContextLogger {
+  logger = null;
+  handlers = [];
+  args = [];
+
+  constructor(logger, handlers, args) {
+    this.logger = logger;
+    this.handlers = handlers;
+    this.args = args;
+  }
+
+  _contextualData(data) {
+    return _.extend(handlerData(this.handlers, ...this.args), data);
+  }
+
+  infoD(title, data) {
+    this.logger.infoD(title, this._contextualData(data));
+  }
```

Comment:
Definitely want to add all methods. Not sure if we can loop to add functions to a class though.
Anyways, I started `infoD` as a POC.